framework/include/utils/RankFourTensorImplementation.h

```diff
@@ -0,0 +1,902 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "RankFourTensor.h"
+
+// MOOSE includes
+#include "RankTwoTensor.h"
+#include "MooseEnum.h"
+#include "MooseException.h"
+#include "MooseUtils.h"
+#include "MatrixTools.h"
+#include "PermutationTensor.h"
+
+#include "libmesh/utility.h"
+#include "libmesh/tensor_value.h"
+#include "libmesh/vector_value.h"
+
+// C++ includes
+#include <iomanip>
+#include <ostream>
+
+namespace MathUtils
+{
+template <>
+void
+mooseSetToZero<RankFourTensorTempl<Real>>(RankFourTensorTempl<Real> & v)
```

Comment:
As a full template specialization, won't this cause ODR failures if this header ever gets included in multiple translation units? Seems like maybe the best thing to do would be to move this definition into `RankFourTensor.C`? Same for the `DualReal` version as well as analogs in `RankThreeTensor` and `RankTwoTensor`.

Reply:
But... we have `#pragma once` :-)

Reply:
> multiple translation units?

is the key here. Unless you're campaigning for a `#pragma program_once`? 😄

Reply:
Oh, not in the implementation. Fixing it properly.